### PR TITLE
[interp] Intrinsify span ctor

### PR DESCRIPTION
**Pull request by BrzVlad**

### CODE REVIEW

File: src/mono/mono/mini/interp/transform.c [hunk 1]
```diff
@@ -4743,6 +4752,14 @@ generate_code (TransformData *td, MonoMethod *method, MonoMethodHeader *header,
 					/* public ByReference(ref T value) */
 					g_assert (csignature->hasthis && csignature->param_count == 1);
 					interp_add_ins (td, MINT_INTRINS_BYREFERENCE_CTOR);
+				} else if (m_class_get_image (klass) == mono_defaults.corlib &&
+						(!strcmp (m_class_get_name (m->klass), "Span`1") ||
+						!strcmp (m_class_get_name (m->klass), "ReadOnlySpan`1")) &&
+						csignature->param_count == 2 &&
+						csignature->params [0]->type == MONO_TYPE_PTR &&
+						!type_has_references (mono_method_get_context (m)->class_inst->type_argv [0])) {
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:4760]: @vargaz type_has_references is hackish. Do we have some api in the runtime to check if a type has references, or to get the class associated with a type ?

Comment (vargaz) [reply src/mono/mono/mini/interp/transform.c:4760]: mono_class_from_mono_type ().

**Review by vargaz — APPROVED** [review]

File: src/mono/mono/mini/interp/transform.c (hunk 1 above)
Comment (BrzVlad) [reply src/mono/mono/mini/interp/transform.c:4759]: not sure how I couldn't find it since I even used it before in interp

**Comment by SamMonoRT** [comment]
What kind of perf gain are we looking at with this change ?

**Comment by steveharter** [comment]
I assume this would also help methods like `Slice()` since that returns a new Span.

**Comment by steveharter** [comment]
Here's a simple benchmark doing `new Span<byte>(bytes).Slice()` (100,000,000 iterations)
| Core JIT | Mono JIT | Mono interpreter |
| :-- | :-- | :-- |
| 23ms | 652 (28x slower than Core) | 7597 (330x slower than Core)

**Pull request merged**